### DISCUSSION TITLE
Thoughts on .Setup, .Calls

### DISCUSSION THREAD

**Discussion by jsk95** [ashmind/SourceMock#5] [Ideas]
```
interface IParser { int Parse(string value); }

var parser = Mock.Of<IParser>().Get();
parser.Setup.Parse().Returns(1);

Assert.Equal(1, parser.Parse());
```

to

```
interface IParser { int Parse(string value); }

var (parser, parserSetup, _) = Mock.Of<IParser>().Get();
parserSetup.Parse().Returns(1);

Assert.Equal(1, parser.Parse());
```

**Comment by jsk95** [comment]
Hm actually might be annoying with traditional test setups in C# atm

**Comment by ashmind** [comment]
What is the main benefit in this approach?
As in, you will have to deconstruct _always_, as you wouldn't need a mock without setup or calls.

This also limits reusable code a bit, e.g. you have to return setup object if others wants to set up one more thing.
